tests/cover/test_regex.py

```diff
@@ -0,0 +1,396 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+import hypothesis as h
+import hypothesis.errors as he
+from hypothesis.internal.compat import PY3, hrange, hunichr
+
+from hypothesis.searchstrategy.regex import strategy as regex, \
+    UNICODE_CATEGORIES, UNICODE_DIGIT_CATEGORIES, \
+    UNICODE_SPACE_CATEGORIES, UNICODE_WORD_CATEGORIES, UNICODE_WEIRD_NONWORD_CHARS, \
+    SPACE_CHARS, UNICODE_SPACE_CHARS, HAS_WEIRD_WORD_CHARS
+import pytest
+import re
+import sys
+import unicodedata
+
+
+def is_ascii(s):
+    return all(ord(c) < 128 for c in s)
+
+
+def is_digit(s):
+    return all(unicodedata.category(c) in UNICODE_DIGIT_CATEGORIES for c in s)
+
+
+def is_space(s):
+    return all(c in SPACE_CHARS for c in s)
+
+
+def is_unicode_space(s):
+    return all(
+        unicodedata.category(c) in UNICODE_SPACE_CATEGORIES or \
+        c in UNICODE_SPACE_CHARS
+        for c in s
+    )
+
+
+def is_word(s):
+    return all(
+        c == '_' or (
+            (not HAS_WEIRD_WORD_CHARS or c not in UNICODE_WEIRD_NONWORD_CHARS) and
+            unicodedata.category(c) in UNICODE_WORD_CATEGORIES
+        )
+        for c in s
+    )
+
+
+def ascii_regex(pattern):
+    flags = re.ASCII if PY3 else 0
+    return re.compile(pattern, flags)
+
+
+def unicode_regex(pattern):
+    return re.compile(pattern, re.UNICODE)
+
+
+def _test_matching_pattern(pattern, isvalidchar, unicode=False):
```

Comment:
nit: use an argument name that doesn't shadow the `unicode` builtin on Python 2.